pkg/agent/ipassigner/linkmonitor/link_monitor_linux.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2024 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package linkmonitor
+
+import (
+	"sync"
+	"time"
+
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/set"
+
+	utilnetlink "antrea.io/antrea/pkg/agent/util/netlink"
+)
+
+const (
+	linkAny = ""
+)
+
+type linkMonitor struct {
+	mutex             sync.RWMutex
+	cacheSynced       bool
+	linkSubscribeFunc func(ch chan<- netlink.LinkUpdate, done <-chan struct{}, options netlink.LinkSubscribeOptions) error
+	eventHandlers     map[string][]LinkEventHandler
+	linkNames         set.Set[string]  // known link names
+	linkIndexMap      map[int32]string // map from link index to link name
+	netlink           utilnetlink.Interface
+}
+
+func NewLinkMonitor() *linkMonitor {
+	return &linkMonitor{
+		linkSubscribeFunc: netlink.LinkSubscribeWithOptions,
+		eventHandlers:     make(map[string][]LinkEventHandler),
+		linkNames:         set.New[string](),
+		linkIndexMap:      make(map[int32]string),
+		netlink:           &netlink.Handle{},
+	}
+}
+
+func (d *linkMonitor) HasSynced() bool {
+	d.mutex.RLock()
+	defer d.mutex.RUnlock()
+	return d.cacheSynced
+}
+
+func (d *linkMonitor) AddEventHandler(handler LinkEventHandler, linkNames ...string) {
+	if len(linkNames) == 0 {
+		d.eventHandlers[linkAny] = append(d.eventHandlers[linkAny], handler)
+		return
+	}
+	for _, name := range linkNames {
+		d.eventHandlers[name] = append(d.eventHandlers[name], handler)
+	}
+}
+
+func (d *linkMonitor) Run(stopCh <-chan struct{}) {
+	klog.InfoS("Starting localLinkMonitor")
+
+	go wait.NonSlidingUntil(func() {
+		d.listAndWatchLinks(stopCh)
+	}, 5*time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (d *linkMonitor) listAndWatchLinks(stopCh <-chan struct{}) {
+	ch := make(chan netlink.LinkUpdate)
```

Comment:
is an unbuffered channel desired here?

cc @tnqn

Reply:
I had the question in mind too, but found I also used unbuffered channel in ip detector. I don't quite remember why I did it but it seems working fine so I didn't raise the question.

But I do agree a buffered comment is more suitable in both detectors. We don't need to make the producer and consumer synchornous.

Reply:
changed to buffered channel (capacity is 1).

Reply:
I think the buffer could be greater, like 100. Capacity 1 is usually used on purpose in some scenario, here we want to prevent the producer from being blocked, so a greater capaicity makes more sense.

Reply:
ok. changed to 100.